wpilibj/src/athena/java/edu/wpi/first/wpilibj/DriverStation.java

```diff
@@ -636,18 +643,43 @@ public void waitForData() {
    * Wait for new data or for timeout, which ever comes first. If timeout is 0, wait for new data
    * only.
    *
-   * @param timeout The maximum time in milliseconds to wait.
+   * @param timeout The maximum time in seconds to wait.
+   * @return true if there is new data, otherwise false
    */
-  public void waitForData(long timeout) {
-    synchronized (m_dataSem) {
+  public boolean waitForData(double timeout) {
+    long startTime = Utility.getFPGATime();
+    long timeoutMicros = (long)(timeout * 1000000);
+    m_dataMutex.lock();
+    try {
       try {
         while (!m_updatedControlLoopData) {
-          m_dataSem.wait(timeout);
+          if (timeout > 0) {
+            long now = Utility.getFPGATime();
+            if (now < startTime + timeoutMicros) {
+              // We still have time to wait
+              boolean signaled = m_dataCond.await(startTime + timeoutMicros - now, 
+                                                  TimeUnit.MICROSECONDS);
+              if (!signaled) {
+                // Return false if a timeout happened
+                return false;
+              }
+            } else {
+              // Time has elapsed.
+              return false;
+            }
+          } else {
+            m_dataCond.await();
+          }
         }
         m_updatedControlLoopData = false;
+        // Return true if we have received a proper signal
+        return true;
       } catch (InterruptedException ex) {
-        Thread.currentThread().interrupt();
+        // return false on a thread interrupt
+        return false;
```

Comment:
I'm not an expert on InterruptedException, but is this the right thing to do here?